File: android/audio-contracts/src/main/kotlin/com/calliopeia/edgeaudio/contracts/AudioTypes.kt
package com.calliopeia.edgeaudio.contracts

enum class CaptureMode {
    RAW_MASTER,
    SYSTEM_VOICE,
}

enum class EnhancementProfile {
    BYPASS,
    ADAPTIVE_DSP,
    GTCRN_16K,
    DPDFNET_16K,
    DPDFNET_48K,
}

data class AudioFrame(
    val samples: FloatArray,
    val sampleRate: Int,
    val timestampMilliseconds: Long,
)

enum class QualityFlag {
    SILENCE,
    TOO_QUIET,
    CLIPPING,
    LOW_SNR,
}

data class QualitySnapshot(
    val rmsDbFS: Double,
    val peakDbFS: Double,
    val noiseFloorDbFS: Double,
    val estimatedSnrDb: Double,
    val clippingRatio: Double,
    val flags: Set<QualityFlag>,
)

interface AudioFrameInspector {
    fun reset()
    fun inspect(samples: FloatArray): QualitySnapshot
}

data class CaptureFormat(
    val requestedSampleRate: Int,
    val actualSampleRate: Int,
    val channelCount: Int,
    val isFloatPcm: Boolean,
    val mode: CaptureMode,
    val audioSource: Int,
)

interface StreamingAudioEnhancer : AutoCloseable {
    val requiredSampleRate: Int
    val preferredFrameSize: Int
    fun process(samples: FloatArray, sampleRate: Int): FloatArray
    fun flush(): FloatArray = floatArrayOf()
    fun reset()
    override fun close() = Unit
}

class UnsupportedSampleRateException(expected: Int, actual: Int) :
    IllegalArgumentException("Expected ${expected} Hz but received ${actual} Hz")

File: ios/Sources/CalliopeiaAudioContracts/AudioTypes.swift
import Foundation

public enum CaptureMode: String, Codable, Sendable {
    case rawMaster
    case systemVoice
}

public enum EnhancementProfile: String, Codable, Sendable {
    case bypass
    case adaptiveDsp
    case gtcrn16k
    case dpdfnet16k
    case dpdfnet48k
}

public struct AudioFrame: Sendable {
    public let samples: [Float]
    public let sampleRate: Int
    public let timestampMilliseconds: Int64

    public init(samples: [Float], sampleRate: Int, timestampMilliseconds: Int64) {
        self.samples = samples
        self.sampleRate = sampleRate
        self.timestampMilliseconds = timestampMilliseconds
    }
}

public enum QualityFlag: String, Codable, Hashable, Sendable {
    case silence
    case tooQuiet
    case clipping
    case lowSnr
}

public struct QualitySnapshot: Codable, Equatable, Sendable {
    public let rmsDbFS: Double
    public let peakDbFS: Double
    public let noiseFloorDbFS: Double
    public let estimatedSnrDb: Double
    public let clippingRatio: Double
    public let flags: Set<QualityFlag>

    public init(
        rmsDbFS: Double,
        peakDbFS: Double,
        noiseFloorDbFS: Double,
        estimatedSnrDb: Double,
        clippingRatio: Double,
        flags: Set<QualityFlag>
    ) {
        self.rmsDbFS = rmsDbFS
        self.peakDbFS = peakDbFS
        self.noiseFloorDbFS = noiseFloorDbFS
        self.estimatedSnrDb = estimatedSnrDb
        self.clippingRatio = clippingRatio
        self.flags = flags
    }
}

public protocol AudioFrameInspecting: AnyObject, Sendable {
    func reset()
    func inspect(samples: [Float]) -> QualitySnapshot
}

public struct CaptureFormat: Codable, Equatable, Sendable {
    public let requestedSampleRate: Int
    public let actualSampleRate: Int
    public let channelCount: Int
    public let isFloatPCM: Bool
    public let mode: CaptureMode

    public init(
        requestedSampleRate: Int,
        actualSampleRate: Int,
        channelCount: Int,
        isFloatPCM: Bool,
        mode: CaptureMode
    ) {
        self.requestedSampleRate = requestedSampleRate
        self.actualSampleRate = actualSampleRate
        self.channelCount = channelCount
        self.isFloatPCM = isFloatPCM
        self.mode = mode
    }
}

public protocol StreamingAudioEnhancer: AnyObject {
    var identifier: String { get }
    var requiredSampleRate: Int { get }
    var preferredFrameSize: Int { get }
    func process(samples: [Float], sampleRate: Int) throws -> [Float]
    func flush() throws -> [Float]
    func reset()
}

public extension StreamingAudioEnhancer {
    var identifier: String { String(reflecting: type(of: self)) }
}

public enum AudioEnhancerError: Error, Equatable {
    case unsupportedSampleRate(expected: Int, actual: Int)
    case invalidModel(String)
    case processingFailed
}
